desc/objectives/_free_boundary.py

```diff
@@ -0,0 +1,840 @@
+                self._sheet_data_keys, obj=sheet_current, grid=eval_grid
+            )
+            sheet_src_transforms = get_transforms(
+                self._sheet_data_keys, obj=sheet_current, grid=src_grid
+            )
+            self._constants["sheet_eval_transforms"] = sheet_eval_transforms
+            self._constants["sheet_src_transforms"] = sheet_src_transforms
+
+        timer.stop("Precomputing transforms")
+        if verbose > 1:
+            timer.disp("Precomputing transforms")
+
+        self._dim_f = neq * eval_grid.num_nodes
+
+        if self._normalize:
+            scales = compute_scaling_factors(eq)
+            self._normalization = scales["B"] * scales["R0"] * scales["a"]
+
+        super().build(use_jit=use_jit, verbose=verbose)
+
+    def compute(self, eq_params, sheet_params=None, constants=None):
+        """Compute boundary force error.
+
+        Parameters
+        ----------
+        eq_params : dict
+            Dictionary of equilibrium degrees of freedom, eg Equilibrium.params_dict
+        sheet_params : dict
+            Dictionary of sheet current degrees of freedom.
+        constants : dict
+            Dictionary of constant data, eg transforms, profiles etc. Defaults to
+            self.constants
+
+        Returns
+        -------
+        f : ndarray
+            Boundary force error (N).
+
+        """
+        if constants is None:
+            constants = self.constants
+        src_data = compute_fun(
+            "desc.equilibrium.equilibrium.Equilibrium",
+            self._eq_data_keys,
+            params=eq_params,
+            transforms=constants["src_transforms"],
+            profiles=constants["src_profiles"],
+        )
+        eval_data = compute_fun(
+            "desc.equilibrium.equilibrium.Equilibrium",
+            self._eq_data_keys,
+            params=eq_params,
+            transforms=constants["eval_transforms"],
+            profiles=constants["eval_profiles"],
+        )
+        if self._sheet_current:
+            assert sheet_params is not None
+            # enforce that they're the same surface
+            sheet_params["R_lmn"] = eq_params["Rb_lmn"]
+            sheet_params["Z_lmn"] = eq_params["Zb_lmn"]
+            sheet_src_data = compute_fun(
+                "desc.magnetic_fields.FourierCurrentPotentialField",
+                self._sheet_data_keys,
+                params=sheet_params,
+                transforms=constants["sheet_src_transforms"],
+                profiles={},
+            )
+            sheet_eval_data = compute_fun(
+                "desc.magnetic_fields.FourierCurrentPotentialField",
+                self._sheet_data_keys,
+                params=sheet_params,
+                transforms=constants["sheet_eval_transforms"],
+                profiles={},
+            )
+            src_data["K_vc"] += sheet_src_data["K"]
+
+        # this is in cartesian
+        Bplasma = -virtual_casing_biot_savart(
```

Comment:
reason for the negative sign?

I'd like the sign conventions to be clearly laid out at least in the objective so it is clear what is what i.e. that K = n x B or B x n, n is an outward facing normal, etc

Reply:
I was using the opposite sign convention to the current potential stuff, it should be fixed now. I've added a more complete explanation of the virtual casing to the docsting of that function, for the objective I'm just focusing on what equations its actually minimizing rather than how its calculated.